feat: new page overviewing CI & code coverage

lwasser opened the pull request:
This pr is the final piece of content from the now closed pr #104 that was opened in september. it contains 2 short pages

1. describing CI
2. overviewing code coverage. 

because it's been so long i plan to review / edit prior to making this reviewable.

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1206992329524474

lwasser commented on conf.py:
```diff
@@ -97,7 +97,7 @@
         "image_light": "logo-light-mode.png",
         "alt_text": "pyOpenSci Python Package Guide. The pyOpenSci logo is a purple flower with pyOpenSci under it. The o in open sci is the center of the flower",
     },
-    "header_links_before_dropdown": 4,
```
This makes space for a CI&data section in our navigation.

willingc commented:
@lwasser Should we close this draft? Or do you want to leave open.

lwasser commented:
@willingc i think we need to keep this one open. its content that i have yet to finish. so maybe a help-wanted tag is warranted. this page isn't high priority now but i can try to put it in my task list as a chip away at it task.

lwasser commented:
tasks

- [ ] rebase against main
- [ ] revisit content and figure out what section it belongs in
- [ ] edit 
- [ ] send through community review

ucodery commented on ci-data/ci.md:
```diff
@@ -0,0 +1,66 @@
+- Running your test suite
+- Running code checkers / linters / spellcheck
+- Building your documentation
+- Deploying your documentation to GitHub pages
```
You also note this as a use-case of CD.
```suggestion

```

blink1073 reviewed (approved):
LGTM!

tkoyama010 commented:
pre-commit.ci autofix

sneakers-the-rat commented:
i'm a little confused about why CI and data are combined? it seems worth it to me to have a dedicated `ci` section and move all ci stuff there (e.g. i would probably also move the code coverage part to `ci` as well, because one typically wants coverage in a CI context vs. in a general testing context) and `data` feels more like something that would go in the `tests` section: "how do i handle bulky data i need for my tests." or if we were to include it in `ci` it should just be a section within `ci` rather than the section being called `ci-data`

webknjaz commented on tests/code-cov.md:
```diff
@@ -0,0 +1,13 @@
+:height: 450px
+:alt: Screenshot of the code cov service - showing test coverage for the stravalib package. in this image you can see a list of package modules and the associated number of lines and % lines covered by tests. at the top of the image you can see what branch is being evaluated and the path to the repository being shown.
+
+The CodeCov platform is a useful tool if you wish to visually track code coverage. Using it you can not only get the same summary information that you can get with **pytest-cov** extension. You can also get a visual representation of what lines are covered by your tests and what lines are not covered. Code coverage is mostly useful for evaluating unit tests and/or how much of your package code is ...
```
It may be useful to mention that MyPy is also able to produce coverage reports for typing that can be uploaded to Codecov and overlaid on top of runtime testing coverage.

Here's one place where I implemented this:
* https://github.com/aio-libs/yarl/blob/3967268/.github/workflows/reusable-linters.yml#L66C11-L69C51
* https://github.com/aio-libs/yarl/blob/3967268/.pre-commit-config.yaml#L127C7-L127C57

It comes with a few challenges, though: https://discuss.python.org/t/is-there-any-tool-that-can-report-type-coverage-of-a-project/34962/4.

Also, there's GitHub Actions like `irongut/CodeCoverageSummary` that can be helpful in presenting this in job summaries: https://github.com/cherrypy/cheroot/blob/a9c1fe4/.github/workflows/ci-cd.yml#L1100C5-L1112C65.

Finally, Coveragepy itself is able to output markdown for use in job summaries (https://github.com/aio-libs/yarl/blob/3967268/.github/workflows/ci-cd.yml#L331C9-L332C36) and it can produce HTML output locally.
The latter is quite useful since it will display context details that are not represented on Codecov. Plus, it's nice not to have to push to remote and trigger an expensive CI run to work on something locally and maybe even offline.

lwasser replied:
thank you!! this is super helpful. I've updated the file to add info about Mypy and exporting results in a local or ci build to avoid needing to use the website interfaces! !!

lwasser commented:
thank you, everyone, for the reviews!! Once Ci builds and is happy i'll merge this!!

lwasser commented:
@all-contributors please add @webknjaz for review

lwasser commented:
@all-contributors add @namurphy for code, review

lwasser commented:
Ok, I've now renamed the CI section continuous-integration.
I've removed data from the section title. We can consider putting David's PR #110 in the tests section when we review it in 2025 as it fits there. 

The ci section is quite thin, but the one page we have, I think, is a great start 🚀  thank you, everyone. This is getting merged now!! 🎉
